zcash_client_memory/src/mem_wallet/tables.rs

```diff
@@ -0,0 +1,370 @@
+#![allow(unused)]
+use core::time;
+use incrementalmerkletree::{Address, Marking, Position, Retention};
+use sapling::NullifierDerivingKey;
+use secrecy::{ExposeSecret, SecretVec};
+use shardtree::{error::ShardTreeError, store::memory::MemoryShardStore, ShardTree};
+use std::{
+    cmp::Ordering,
+    collections::{hash_map::Entry, BTreeMap, BTreeSet, HashMap, HashSet},
+    convert::Infallible,
+    hash::Hash,
+    num::NonZeroU32,
+    ops::Deref,
+};
+use zcash_keys::keys::{AddressGenerationError, DerivationError, UnifiedIncomingViewingKey};
+use zip32::{fingerprint::SeedFingerprint, DiversifierIndex, Scope};
+
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::{BlockHeight, Network},
+    transaction::{components::OutPoint, txid, Authorized, Transaction, TransactionData, TxId},
+};
+use zcash_protocol::{
+    memo::{self, Memo, MemoBytes},
+    value::{ZatBalance, Zatoshis},
+    PoolType,
+    ShieldedProtocol::{self, Orchard, Sapling},
+};
+
+use zcash_client_backend::{
+    address::UnifiedAddress,
+    data_api::{
+        chain::ChainState, Account as _, AccountPurpose, AccountSource, SeedRelevance,
+        TransactionDataRequest, TransactionStatus,
+    },
+    keys::{UnifiedAddressRequest, UnifiedFullViewingKey, UnifiedSpendingKey},
+    wallet::{Note, NoteId, WalletSaplingOutput, WalletSpend, WalletTransparentOutput, WalletTx},
+};
+
+use zcash_client_backend::data_api::{
+    chain::CommitmentTreeRoot, scanning::ScanRange, AccountBirthday, BlockMetadata,
+    DecryptedTransaction, NullifierQuery, ScannedBlock, SentTransaction, WalletCommitmentTrees,
+    WalletRead, WalletSummary, WalletWrite, SAPLING_SHARD_HEIGHT,
+};
+
+use super::AccountId;
+
+#[cfg(feature = "transparent-inputs")]
+use {
+    zcash_client_backend::wallet::TransparentAddressMetadata,
+    zcash_primitives::legacy::TransparentAddress,
+};
+
+#[cfg(feature = "orchard")]
+use {
+    zcash_client_backend::data_api::ORCHARD_SHARD_HEIGHT,
+    zcash_client_backend::wallet::WalletOrchardOutput,
+};
+
+use crate::error::Error;
+
+/// Maps a block height and transaction index to a transaction ID.
+pub struct TxLocatorMap(HashMap<(BlockHeight, u32), TxId>);
+
+/// Maps a block height and transaction (i.e. transaction locator) index to a nullifier.
+pub struct NullifierMap(BTreeMap<Nullifier, (BlockHeight, u32)>);
+
+/// Keeps track of notes that are spent in which transaction
+pub struct ReceievdNoteSpends(HashMap<NoteId, TxId>);
+
+pub struct ReceivedNoteTable(pub Vec<ReceivedNote>);
```

Comment:
We could encapsulate this if we can

```suggestion
pub struct ReceivedNoteTable(Vec<ReceivedNote>);
```